### PR TITLE
Always create objects that have no prototype

### PR DESCRIPTION
**Pull request by sabberworm**
Fixes vulnerability that allows injecting properties into Object.prototype.

Compare result of `deparam('test[__proto__][test]=1')`.

Previously:

```javascript
deparam('test[__proto__][test]=1'); // evaluates to { test: {} }
{}.test // evaluates to 1
```

Now:
```javascript
deparam('test[__proto__][test]=1'); // evaluates to {test:{__proto__:{test:1}}}
{}.test // evaluates to undefined
```

Drawbacks:

* `Object.create` isn’t supported in older IEs
* Debuggers print the returned objects a bit differently (e.g. `[Object: null prototype] {}` instead of `{}`
* Prototype properties and methods don’t exist on the returned objects (could be a backwards-incompatible change)

Alternate solution would be to filter the object keys more rigorously…

### CODE REVIEW

File: src/deparam.js [hunk 1]
```diff
@@ -52,7 +52,7 @@ function deparam(qs, coerce) {
     coerce = setDefault(coerce, true);
     qs = qs.substring(qs.charAt(0) === '?');
     const queryParamList = qs.split('&');
-    const queryObject = {};
+    const queryObject = Object.create(null);
```
Comment (sabberworm) [inline src/deparam.js]: @scssyworks was there a reason to revert this?

File: src/deparam.js [hunk 2]
```diff
@@ -121,7 +121,7 @@ function complex(key, value, obj, doCoerce) {
         } else if (nextProp) {
             const { ob, push } = resolveObj(obj[prop], nextProp);
             obj[prop] = ob;
-            const nextOb = push ? {} : obj[prop];
+            const nextOb = push ? Object.create(null) : obj[prop];
```
Comment (sabberworm) [inline src/deparam.js]: @scssyworks this one you also reverted. I don’t understand why. But I also don’t understand whether it makes a difference or not…

File: src/deparam.js (hunk 1 above)
Comment (sabberworm) [reply src/deparam.js]: At least this one makes a difference:

```javascript
deparam('__proto__[a]=b'); // Will pollute prototype
```

File: src/deparam.js (hunk 2 above)
Comment (sabberworm) [reply src/deparam.js]: This is also vulnerable:

```javascript
deparam('test[a]=b&test[a][bla]=c&test[a][1][__proto__][test]=yay'); // Will put test: yay on the prototype
```

File: src/deparam.js (hunk 1 above)
Comment (scssyworks) [reply src/deparam.js]: @sabberworm Apologies! Seems like it's a mistake on my part. I believe you have already raised an MR to fix that. Thank you for that!

File: src/deparam.js (hunk 2 above)
Comment (scssyworks) [reply src/deparam.js]: Once again apologies on the merge error

**Pull request merged**